### PR TITLE
Analytics schema v2 - authImpression event

### PR DESCRIPTION
**Pull request by eessex**
An MVP demonstrating helper functions that create tracking args for segment events. This PR creates the `authImpression` event, which strictly types arguments that are allowed for this particular tracking action. 

The v2 analytics directory is separated into two folders:
- `/Schema`, which holds schemas for individual events and property types on events
- `/Events`, which holds helper functions like `authImpression` to generate event data that conforms to the schema.

This work is intended to be published as an NPM package, but will hold on moving this into a dependency until we have proven this event works as expected.

TODO:
- [x] Force auth triggers conform to schema
- [x] Determine downstream migration steps for Redshift value changes

Related context: https://www.notion.so/Analytics-Schema-Improvements-4e091c7186c8417b9f9fa05690570834

### CODE REVIEW

File: src/Artsy/Analytics/v2/Events/AuthImpression.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+interface AuthImpressionArgs {
+  copy?: string
+  contextModule: AuthContextModule
+  intent: AuthIntent
+  onboarding?: boolean
+  trigger?: AuthTrigger
+  triggerSeconds?: number
+  type: AuthModalType
+}
```
Comment (cavvia) [inline src/Artsy/Analytics/v2/Events/AuthImpression.ts]: Is there any way of avoiding some of the repetition here? The `AuthImpression` interface and the args declaration are very similar.

File: src/Artsy/Analytics/v2/Events/AuthImpression.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  modal_copy?: string
+  onboarding: boolean
+  trigger: AuthTrigger
+  trigger_seconds?: number
+  type: AuthModalType
+}
+
+interface AuthImpressionArgs {
+  copy?: string
+  contextModule: AuthContextModule
```
Comment (cavvia) [inline src/Artsy/Analytics/v2/Events/AuthImpression.ts]: Good to see strict typing on standard properties like these :+1:

File: src/Artsy/Analytics/v2/Schema/ContextModule.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  categoryRail = "categoryRail",
+  featuredArtistsRail = "featuredArtistsRail",
+  header = "header",
+  otherWorksByArtistRail = "otherWorksByArtistRail",
+  otherWorksByGalleryRail = "otherWorksByGalleryRail",
+  relatedArtistsRail = "relatedArtistsRail",
+  relatedWorksRail = "relatedWorksRail",
+  worksByPopularArtistsRail = "worksByPopularArtistsRail",
+  worksForSaleRail = "worksForSaleRail",
+}
```
Comment (cavvia) [inline src/Artsy/Analytics/v2/Schema/ContextModule.ts]: :+1:

File: src/Artsy/Analytics/v2/Events/AuthImpression.ts (hunk 1 above)
Comment (eessex) [reply src/Artsy/Analytics/v2/Events/AuthImpression.ts]: We gain two things from using a different interface for args - some options can be optional and have default values, and the naming convention. In JS apps our convention uses `contextModule` for args, and this format is ideal. Segment expects `context_module`, so the keys are different on the `AuthImpression` event interface.

**Comment by cavvia** [comment]
Thanks for this! I like that there is stricter typing here than is currently the case. Definitely a step in the right direction.

My general sense is that there is a fair amount of boilerplate here for the declaration of a single event. I wonder if we can streamline this somehow. I would like analysts to be able to declare events without having to understand typescript, for example. In this context I've mentioned JSON Schema in the past. 

Another concern is that the schema may potentially be spread out over many files under this scheme. e.g. adding a new context module to an `AuthImpression` event requires an update in two places here. Is there any way to streamline this by centralizing the schema a bit more? Having more of an overview of the schema is pretty useful to analysts.

File: src/Apps/Artist/Routes/Overview/Components/ZeroState.tsx [hunk 4]
```diff
@@ -1,8 +1,9 @@
 import { Link, Message } from "@artsy/palette"
-import { AnalyticsSchema, useSystemContext } from "Artsy"
```
Comment (damassi) [inline src/Apps/Artist/Routes/Overview/Components/ZeroState.tsx:2]: @eessex - we should avoid `import * as SchemaV2` and stick to the named export format as seen here

File: src/Apps/Artist/Routes/Overview/Components/WorksForSaleRail.tsx [hunk 5]
```diff
@@ -39,6 +40,7 @@ const WorksForSaleRail: React.FC<WorksForSaleRailProps & {
         return (
           <FillwidthItem
             artwork={artwork.node}
+            contextModule={ContextModule.worksForSaleRail}
```
Comment (damassi) [inline src/Apps/Artist/Routes/Overview/Components/WorksForSaleRail.tsx:43]:
```js
import { AnalyticsSchemaV2 } from "Artsy"
...
contextModule={AnalyticsSchemaV2.ContextModule.worksForSaleRail}
```
that way we have a consistent way to find and replace in the future aligned with the rest of our analytics imports.

File: src/Apps/Artwork/Components/ArtworkImageBrowser/ArtworkActions.tsx [hunk 6]
```diff
@@ -31,6 +31,7 @@ import {
   ShareIcon,
   Spacer,
 } from "@artsy/palette"
+import { ContextModule } from "Artsy/Analytics/v2/Schema"
```
Comment (damassi) [inline src/Apps/Artwork/Components/ArtworkImageBrowser/ArtworkActions.tsx:34]: Same here

File: src/Apps/Artwork/Components/OtherWorks/RelatedWorksArtworkGrid.tsx [hunk 7]
```diff
@@ -13,6 +13,7 @@ import React, { useContext } from "react"
 import styled from "styled-components"
 import createLogger from "Utils/logger"
 
+import { ContextModule } from "Artsy/Analytics/v2/Schema"
```
Comment (damassi) [inline src/Apps/Artwork/Components/OtherWorks/RelatedWorksArtworkGrid.tsx:16]: (And all further imports)

File: src/Artsy/Analytics/v2/index.ts [hunk 8]
```diff
@@ -0,0 +1,5 @@
+import * as Schema from "./Schema"
+export const AnalyticsSchema = Schema
```
Comment (damassi) [inline src/Artsy/Analytics/v2/index.ts:2]: export const AnalyticsSchemaV2 = Schema

**Comment by eessex** [comment]
I will be moving this to an NPM package next week once this is proved e2e - @damassi your concerns of imports are noted, but I think it will make more sense to move them when we update all the imports are rerouted to the package.

File: src/Components/FollowButton/FollowArtistButton.tsx [hunk 9]
```diff
@@ -77,17 +78,22 @@ export class FollowArtistButton extends React.Component<Props, State> {
...
+    const trackingData: FollowTrackingData = this.props.trackingData || {}
 
     if (user && user.id) {
       this.followArtistForUser(user)
     } else if (onOpenAuthModal) {
-      const config = {
-        contextModule: "intext tooltip",
-        intent: "follow artist",
+      onOpenAuthModal(ModalType.signup, {
+        contextModule: SchemaV2.ContextModule.intextTooltip,
```
Comment (damassi) [inline src/Components/FollowButton/FollowArtistButton.tsx:88]: 👍

**Comment by damassi** [comment]
Sounds good 👍

**Comment by artsyit** [comment]
:rocket: PR was released in `v26.0.0` :rocket:

**Pull request merged**